File: 2013 letni/Lodz B/ALGO-Lodz-B/src/san/algo/math/Heron.java
package san.algo.math;

public class Heron {

  public static double sqrt(double x) {
    return sqrt(1, x);
  }

  private static double sqrt(double y, double x) {
    if (isEnough(y, x)) {
      return y;
    }
    return sqrt(better(y, x), x);
  }

  private static double better(double y, double x) {
    return (y + x / y) / 2.0;
  }

  private static boolean isEnough(double y, double x) {
    return Math.abs(y * y - x) <= 0.000001;
  }

}

File: 2021 letni/FP (niestacjonarne)/heron.clj
(defn abs
  [x]
  (if (< x 0)
    (- x)
    x))

(defn square
  [x]
  (* x x))

(defn good-enough?
  [G x]
  (<= (abs (- x (square G))) 0.00000001))

(defn avg
  [x y]
  (/ (+ x y) 2))

(defn improve
  [G x]
  (avg G (/ x G)))

(defn heron-sqrt
  [G x]
  (println "G is" G)
  (if (good-enough? G x)
    G

    ;; otherwise,
    (heron-sqrt (improve G x) x)))

(time (heron-sqrt 1 2))

(-
  (double (heron-sqrt 1 2))
  (Math/sqrt 2))

;; Definicja procedury
#_(defn <procedure-name>
    [<arg-1> <arg-2> ... <arg-n>]
    <body>)

;; Wyrażenie warunkowe
#_(if <predicate>
    <consequent>
    <alternative>)

;; JS ternary-conditional expression
;; <predicate> ? <consequent> : <alternative>;
;; np. let v1 = s1 === 'aaa' ? 2 : 3;
